cmd/query/app/querysvc/adjuster/spaniduniquifier_test.go

```diff
@@ -0,0 +1,108 @@
+// Copyright (c) 2019 The Jaeger Authors.
+// Copyright (c) 2017 Uber Technologies, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+package adjuster
+
+import (
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"go.opentelemetry.io/collector/pdata/pcommon"
+	"go.opentelemetry.io/collector/pdata/ptrace"
+
+	"github.com/jaegertracing/jaeger/internal/jptrace"
+)
+
+var (
+	clientSpanID  = pcommon.SpanID([]byte{0, 0, 0, 0, 0, 0, 0, 1})
+	anotherSpanID = pcommon.SpanID([]byte{1, 0, 0, 0, 0, 0, 0, 0})
+)
+
+func makeTraces() ptrace.Traces {
+	traceID := pcommon.TraceID([]byte{0, 0, 0, 0, 0, 0, 0, 2, 0, 0, 0, 0, 0, 0, 0, 3})
+
+	traces := ptrace.NewTraces()
+	spans := traces.ResourceSpans().AppendEmpty().ScopeSpans().AppendEmpty().Spans()
+
+	clientSpan := spans.AppendEmpty()
+	clientSpan.SetTraceID(traceID)
+	clientSpan.SetSpanID(clientSpanID)
+	clientSpan.SetKind(ptrace.SpanKindClient)
+
+	serverSpan := spans.AppendEmpty()
+	serverSpan.SetTraceID(traceID)
+	serverSpan.SetSpanID(clientSpanID) // shared span ID
+	serverSpan.SetKind(ptrace.SpanKindServer)
+
+	anotherServerSpan := spans.AppendEmpty()
+	anotherServerSpan.SetTraceID(traceID)
+	anotherServerSpan.SetSpanID(clientSpanID) // shared span ID
+	anotherServerSpan.SetKind(ptrace.SpanKindServer)
+
+	anotherSpan := spans.AppendEmpty()
+	anotherSpan.SetTraceID(traceID)
+	anotherSpan.SetSpanID(anotherSpanID)
+
+	return traces
+}
+
+func TestSpanIDUniquifierTriggered(t *testing.T) {
+	traces := makeTraces()
+	deduper := SpanIDUniquifier()
+	require.NoError(t, deduper.Adjust(traces))
+
+	spans := traces.ResourceSpans().At(0).ScopeSpans().At(0).Spans()
+
+	clientSpan := spans.At(0)
+	assert.Equal(t, clientSpanID, clientSpan.SpanID(), "client span ID should not change")
+
+	serverSpan := spans.At(1)
+	assert.EqualValues(t, []byte{0, 0, 0, 0, 0, 0, 0, 2}, serverSpan.SpanID(), "server span ID should be reassigned")
+	assert.EqualValues(t, []byte{0, 0, 0, 0, 0, 0, 0, 3}, serverSpan.ParentSpanID(), "next server span should be this server span's parent")
```

Comment:
I would expect `serverSpan.parentSpanId == clientSpanID`. If it's id=3, there was no such span in the trace before, so it's a broken link.

Reply:
@yurishkuro That's what i was expecting as well but this is a direct translation of what was there before. If you take a look at the [old implementation](https://github.com/jaegertracing/jaeger/blob/main/model/adjuster/zipkin_span_id_uniquify.go#L76), oldToNewSpanIDs gets overwritten with the latest trace that it encounters which is what causes this. Do we want to change the behaviour here?

Reply:
This was due to the undefined behaviour resulting from a having two server spans sharing the same clientID. This is now fixed.